### PR TITLE
Agreement Manager unit tests

### PR DESCRIPTION
**Pull request by daiagi**
1. Unit test suite for `AgreementManager`
2.  Logged  a possible race condition bug: https://trello.com/c/Ij8sVdHd
3. Changes to `AgreementManager`

    - bug fix: `getStateProof` - take participants change set up until requestd block height.
    - change `tryBuildMilestone` from `public` to `private`
    - remove dead methods: `didParticipantPostOnChainLocal`, `getDoubleSignedBlock`
    - move methods `getForkDisputeConfirmations`, `getForkDisputes`, `getReduceData` so that all private methods will be at the bottom of the file

4. Changes to harness
    - bug fix on `submitInvalidStateTransitionBlock`, to allow it to submit the very first block on afork as invalid.
    - additional queries to support unit test suite for `AgreementManager`

5. Single owner for the dispute window
    - `EventSyncService.loadSynchronizedWindowCommitments` - chain read + missing dispute event recovery
    - reduce and spectate both enter through it
    - spectate takes the disputed flag from the chain too, not the local EVM

*

### CODE REVIEW

File: src/agreementManager/AgreementManager.ts [hunk 1]
```diff
@@ -91,9 +91,13 @@ class AgreementManager {
...
         }
 
-        // Get all exit points for this fork
+        // Participant changes up to the requested height only
         const participantChangeHeights =
-            this.storage.participantSetChanges.getChangePointsInRange(forkId);
+            this.storage.participantSetChanges.getChangePointsInRange(
+                forkId,
+                undefined,
+                blockHeight
```
Comment (momoshell) [inline src/agreementManager/AgreementManager.ts:92]: Just a non-blocking nit in it's core, and asking for my understanding.

Is it correct that `getChangePointsInRange(forkId, undefined, blockHeight)` is inclusive at `blockHeight` (`ParticipantSetChangeStorage.ts` filters `height <= actualEnd`)? 

Both new boundary tests ("tops out at requested height") use a `requestedHeight` strictly *below* the change point, so the exact-boundary case (a change landing *at* `blockHeight`) isn't exercised. 

Maybe it's worth a same-height test if the inclusive behavior is intentional?

**Review by momoshell — APPROVED** [review]

**Comment by daiagi** [comment]
rebased on top of #397

**Comment by lukastanisic99** [comment]
Changed the base/target to `dispute` since I merged the rest

**Comment by lukastanisic99** [comment]
image
Flaky here for me - not related to this PR just letting you know

**Comment by lukastanisic99** [comment]
@daiagi long review - mostly take a look at the covers and most of the review is tests that are (potentially missing) - I say potentially since I didn't verify manually across all files. Later this should be easy to verify, since we'll know exactly in which file to look. (Btw, usually the AI can take these review files and just implement the missing tests). 1-2 rounds of ping-pong and they get most of it right and clean. 

## 🧱 Fundamental

- 🔴 **[FR1] `blockHeight` is not enforced as a hard ceiling while building participant-change milestones.** [`getStateProof`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/agreementManager/AgreementManager.ts#L95-L117) filters change points, but its ascending iterator has no end height. [`tryBuildMilestone`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/agreementManager/AgreementManager.ts#L442-L481) can therefore consume later blocks. Targeted sync then aborts when the proof ends above the requested height ([generation](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/rpc/services/spectate/SpectateService.ts#L584-L608), [receiver check](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/rpc/services/spectate/SpectateService.ts#L422-L433)).

  > **Fix FR1-FIX**
  >
  > Bound the ascending iterator to `blockHeight` inclusively. If the change cannot be proved within that bound, return a bounded signed-block proof or no proof; never consume a later block.

## 🔒 Security (trust boundary)

- 🟢 **[SG1] No production trust boundary is widened.** The added RPC surface is test-only, and [`QueryRpcMethods`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/test/fixtures/customRpc/harnessControl/services/query/QueryRpcMethods.ts#L66-L75) keeps helpers and storage access on `QueryService`.

  > **Fix SG1-FIX**
  >
  > No change required.

## 🏁 Race conditions

- 🟠 **[RO1] Spectator sync bypasses the existing recovery for a committed dispute whose local event is pending.** [`generateSyncPayload`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/rpc/services/spectate/SpectateService.ts#L492-L508) calls the throwing local lookup directly, while reduction first replays missing events through [`getSyncedForkDisputes`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/stateManager/reduction/ReductionExecutor.ts#L78-L97).

  > **Fix RO1-FIX**
  >
  > Route spectator dispute-window loading through the same event-recovery owner, then re-read the ordered confirmations before generating the payload.

## ⚡ Performance

- 🟢 **[PG1] The height filter reduces proof-building work.** It avoids scanning participant changes after the requested height and adds no production polling or persistent scans.

  > **Fix PG1-FIX**
  >
  > No change required.

## 🧪 Tests (regressions / gaps)

- 🟠 **[TO1] Unit — `getStateProof` has no exact-height leave case.** The leave test requests height `1`, before the leave ([test](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/test/unit/AgreementManager.test.ts#L233-L248)); it cannot catch a leave milestone consuming later confirmations.

  > **Fix TO1-FIX**
  >
  > Request the exact leave-block height, keep that block partially confirmed, add confirming blocks above it, and assert the result contains no block above the target.

- 🟠 **[TO2] Unit — `getStateProof` has no exact-height join case.** The join test requests height `1`, before the join at height `2` ([test](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/test/unit/AgreementManager.test.ts#L250-L279)); the increased-threshold boundary remains untested.

  > **Fix TO2-FIX**
  >
  > Request the exact join-block height, require later blocks to complete its threshold, and assert the result is bounded to the target or explicitly unavailable.

- 🟠 **[TO3] Unit — reduction's missing-dispute recovery is skipped and asserts obsolete behavior.** The case expects `undefined` ([test](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/test/unit/AgreementManager.test.ts#L648-L742)), while the current owner retries through [`ensureDisputesProcessed`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/stateManager/EventSyncService.ts#L243-L290).

  > **Fix TO3-FIX**
  >
  > Move this variation to a `ReductionExecutor` black-box unit case: remove a committed local dispute, invoke the public reduction path, and assert recovery followed by reduction.

- 🟠 **[TO4] Unit — spectator dispute-window loading does not cover a locally missing committed dispute.** This is the component variation that exposes RO1.

  > **Fix TO4-FIX**
  >
  > Add a `SpectateService` black-box unit case with real storage and event-sync collaborators; suppress local persistence, generate the payload, and assert recovery before confirmation lookup.

- 🟠 **[TO5] E2E — historical targeted sync is not exercised across a participant-set change.** Existing exact-target E2E coverage uses a fixed participant set ([test](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/test/e2e/E2E-Spectate.test.ts#L1071-L1138)).

  > **Fix TO5-FIX**
  >
  > Add one representative spectator workflow: put the responder ahead across a join or leave, request an earlier height, and assert the requester completes sync at exactly that height.

- 🟠 **[TO6] E2E — spectator sync does not prove recovery across the chain-event-storage boundary.** Reduction has representative event-recovery coverage ([existing test](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/test/e2e/E2E-DisputeManager.test.ts#L144-L258)); spectator sync is a different system interaction.

  > **Fix TO6-FIX**
  >
  > Suppress a responder's dispute event while leaving the commitment on-chain, run the real spectator sync flow, and assert recovery produces a valid payload and completed sync.

- 🟡 **[TY7] Unit — `getForkDisputes` lacks a populated-window case.** The only direct case asserts an empty window ([test](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/test/unit/AgreementManager.test.ts#L751-L793)).

  > **Fix TY7-FIX**
  >
  > Store multiple real disputes, request them in commitment order, and assert returned structs and ordering.

- 🟡 **[TY8] Unit — `getForkDisputeConfirmations` lacks a populated-window case.** Broad reduction success does not prove which confirmations this component returns.

  > **Fix TY8-FIX**
  >
  > Commit multiple real disputes and assert confirmation count, encoded disputes, signatures, and commitment order.

- 🟡 **[TY9] Unit — `getForkDisputes` does not cover a missing local dispute.** The public method throws at the first absent commitment ([branch](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/agreementManager/AgreementManager.ts#L360-L374)).

  > **Fix TY9-FIX**
  >
  > Supply one stored and one absent real commitment, then assert the error identifies the absent hash.

- 🟡 **[TY10] Unit — `getForkDisputeConfirmations` does not cover a missing local confirmation.** This branch is distinct from a missing dispute struct ([branch](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/agreementManager/AgreementManager.ts#L339-L357)).

  > **Fix TY10-FIX**
  >
  > Keep an on-chain commitment while removing its local confirmation, then assert the rejection names that commitment.

- 🟡 **[TY11] Unit — `getReduceData` lacks the normal surviving-block path.** The suite exercises only reduction to genesis ([test](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/test/unit/AgreementManager.test.ts#L799-L852)).

  > **Fix TY11-FIX**
  >
  > Reduce to a real surviving block and assert the selected snapshot, encoded state, fork ID, and reduced output.

- 🟡 **[TY12] Unit — `getReduceData` does not cover populated inbound-message bounds and order.** No assertion pins the range passed at [`AgreementManager.ts`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/agreementManager/AgreementManager.ts#L422-L433).

  > **Fix TY12-FIX**
  >
  > Stage message blocks before, inside, and after the reduction range; assert only bounded blocks return in application order.

- 🟡 **[TY13] Unit — `getReduceData` does not cover an empty inbound-message range.** The no-op boundary is separate from selecting a populated range.

  > **Fix TY13-FIX**
  >
  > Use equal lower and upper hashes with no newly applied messages and assert an empty result.

- 🟡 **[TY14] Unit — `getReduceData` does not cover a missing genesis snapshot.** The genesis branch has its own error ([branch](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/agreementManager/AgreementManager.ts#L388-L398)).

  > **Fix TY14-FIX**
  >
  > Remove the fork's genesis snapshot, invoke genesis reduction, and assert the fork-specific error.

- 🟡 **[TY15] Unit — `getReduceData` does not cover a missing surviving-block snapshot.** This is distinct from missing genesis ([branch](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/agreementManager/AgreementManager.ts#L399-L409)).

  > **Fix TY15-FIX**
  >
  > Remove the surviving block's snapshot and assert the missing-latest-snapshot error.

- 🟡 **[TY16] Unit — `getReduceData` does not cover a missing encoded state-machine state.** A valid snapshot can still point at absent local state ([branch](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/agreementManager/AgreementManager.ts#L412-L420)).

  > **Fix TY16-FIX**
  >
  > Keep the selected snapshot, remove its encoded state, and assert the missing-state error.

## ♻️ Code reuse

- 🟠 **[CO1] Dispute-window loading has two owners, and only one performs recovery.** [`AgreementManager.getForkDisputeConfirmations`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/agreementManager/AgreementManager.ts#L339-L357) fetches and maps confirmations, while [`ReductionExecutor.getSyncedForkDisputes`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/src/stateManager/reduction/ReductionExecutor.ts#L78-L97) independently fetches, recovers, and maps disputes.

  > **Fix CO1-FIX**
  >
  > Give synchronized window loading one owner in the event-sync layer, return ordered commitments after recovery, and make reduction and spectator sync derive their local structs from it.

## 📏 AGENTS.md adherence

- 🟠 **[AO1] The PR lacks representative E2E coverage for two affected integrations.** Historical targeted sync across participant changes and spectator dispute-event recovery are separate observable system interactions; TO5 and TO6 cover the gaps.

  > **Fix AO1-FIX**
  >
  > Add the two representative E2E workflows from TO5 and TO6; keep join/leave and exact/before permutations in unit coverage.

- 🟡 **[AY2] Two new primitive collections lack domain meaning.** [`confirmationSignerAddresses: string[]`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/test/fixtures/customRpc/harnessControl/services/query/QueryRpcMethods.ts#L24-L35) lacks an address type or immediate explanation, and [`heightsSeen`](https://github.com/peer3to/state-channels-plus/blob/93498f913c1024e555541dc6e9ad0eb93c288476/test/unit/AgreementManager.test.ts#L289-L303) is `Set<number>` despite the existing `BlockHeight` type.

  > **Fix AY2-FIX**
  >
  > Use `Address[]` and `Set<BlockHeight>`, or add a precise immediate comment where serialization requires primitives.

## 🧹 Miscellaneous (scope)

- 🟢 **[MG1] The remaining changes stay within testability and cleanup scope.** The scripts reuse the existing runner, the harness delegates to production and Solidity verification, and the height-zero helper supports the genesis-reduction case.

  > **Fix MG1-FIX**
  >
  > No change required.

## ❓ Open questions

- 🟢 **[OG1] No product decision is needed before fixing FR1.** The receiver already requires the latest proof block to equal the requested height, so `blockHeight` is a hard inclusive ceiling.

  > **Fix OG1-FIX**
  >
  > Implement and document the existing exact-height contract.

**Comment by lukastanisic99** [comment]
@daiagi no red ones :D 

## 🧱 Fundamental

- 🟢 **[FR1] Addressed: `blockHeight` is now a hard inclusive proof ceiling.** Participant changes are bounded at discovery ([`AgreementManager.ts#L87-L93`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/agreementManager/AgreementManager.ts#L87-L93)), and the ascending milestone walk stops before consuming a higher block ([`AgreementManager.ts#L426-L445`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/agreementManager/AgreementManager.ts#L426-L445)).

  > **Fix FR1-FIX**
  >
  > No change required.

## 🔒 Security (trust boundary)

- 🟠 **[SO1] An authenticated requester can make this state-manager instance replay events for an arbitrary channel.** [`onSpectateRequest`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/rpc/services/spectate/SpectateRpcMethods.ts#L23-L47) passes the remote `syncRequest.channelId` through unchanged. The new path uses it to fetch commitments and query/replay logs ([`SpectateService.ts#L478-L502`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/rpc/services/spectate/SpectateService.ts#L478-L502), [`EventSyncService.ts#L242-L252`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/stateManager/EventSyncService.ts#L242-L252)). Dispatch verifies only that the event matches the requester-selected channel, not `stateManager.channelId` ([`EventSyncService.ts#L342-L371`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/stateManager/EventSyncService.ts#L342-L371)), so foreign-channel events can reach this instance's handlers and storage.

  > **Fix SO1-FIX**
  >
  > Reject a spectate request unless `syncRequest.channelId === stateManager.channelId`. Remove `channelId` from `loadSynchronizedWindowCommitments` and have the service use its owning state manager's channel. Add a crafted-request test proving another channel causes no chain query, replay, or storage mutation.

## 🏁 Race conditions

- 🟢 **[RO1] Addressed: spectator and reduction paths now close the commitment/event/storage gap through the same recovery owner.** Both [`SpectateService`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/rpc/services/spectate/SpectateService.ts#L492-L511) and [`ReductionExecutor`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/stateManager/reduction/ReductionExecutor.ts#L78-L86) load ordered commitments only after `EventSyncService` has recovered missing dispute events.

  > **Fix RO1-FIX**
  >
  > No change required beyond SO1's channel binding.

## ⚡ Performance

- 🟢 **[PG1] The new work is bounded and conditional.** Proof assembly stops at the requested height, and dispute replay returns immediately when every commitment is already local ([`EventSyncService.ts#L255-L263`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/stateManager/EventSyncService.ts#L255-L263)); recovery has a three-attempt bound ([`EventSyncService.ts#L276-L301`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/stateManager/EventSyncService.ts#L276-L301)).

  > **Fix PG1-FIX**
  >
  > No change required.

## 🧪 Tests (regressions / gaps)

- 🟠 **[TO1] Unit — the exact-leave regression does not stage a leave that needs a later block to become final.** The test uses `setupTwoLeaversAcrossMilestones` ([`AgreementManager.test.ts#L280-L319`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/unit/AgreementManager.test.ts#L280-L319)); that helper submits each leave with the default `waitForFinalization` behavior ([`MathScenarioActions.ts#L185-L200`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/harness/actions/math/MathScenarioActions.ts#L185-L200)), which is explicitly `true` ([`MathTransitionActions.ts#L109-L123`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/harness/actions/math/MathTransitionActions.ts#L109-L123)). The old unbounded iterator can therefore pass this test without consuming a higher block.

  > **Fix TO1-FIX**
  >
  > Submit the leave with `waitForFinalization: false`, keep its own block below threshold, add the needed confirmation only above the requested height, and assert the exact-height proof is bounded or unavailable.

- 🟠 **[TO6] E2E — the dispute-event-gap case bypasses the remote spectate request and never completes a spectator sync.** It invokes the responder's test-only `generateSyncPayload` control endpoint directly ([`E2E-Spectate.test.ts#L1368-L1385`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/e2e/E2E-Spectate.test.ts#L1368-L1385)) and expects `null` rather than exercising `onSpectateRequest`, the encoded response, and the requester's receive/apply path ([`E2E-Spectate.test.ts#L1418-L1424`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/e2e/E2E-Spectate.test.ts#L1418-L1424)).

  > **Fix TO6-FIX**
  >
  > Stage a suppressed event on a responder that can still prove the derived target, run the real peer-to-peer spectator request, and assert event recovery, a valid encoded payload, and successful requester application.

- 🟡 **[TY7] Unit — `getForkDisputes` still lacks a populated-window case.** The direct suite covers only empty commitments ([`AgreementManager.test.ts#L804-L860`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/unit/AgreementManager.test.ts#L804-L860)).

  > **Fix TY7-FIX**
  >
  > Store multiple real disputes, request them in commitment order, and assert returned structs and ordering.

- 🟡 **[TY8] Unit — `getForkDisputeConfirmations` still lacks a populated-window case.** Broad reduction success does not pin this mapper's returned confirmations.

  > **Fix TY8-FIX**
  >
  > Commit multiple real disputes and assert confirmation count, encoded disputes, signatures, and commitment order.

- 🟡 **[TY9] Unit — `getForkDisputes` does not cover its missing-local-dispute error.** The branch throws for the first absent commitment ([`AgreementManager.ts#L347-L362`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/agreementManager/AgreementManager.ts#L347-L362)).

  > **Fix TY9-FIX**
  >
  > Supply one stored and one absent real commitment and assert the error identifies the absent hash.

- 🟡 **[TY10] Unit — `getForkDisputeConfirmations` does not cover its missing-local-confirmation error.** This branch is separate from an absent dispute struct ([`AgreementManager.ts#L332-L345`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/agreementManager/AgreementManager.ts#L332-L345)).

  > **Fix TY10-FIX**
  >
  > Keep a commitment while removing its local confirmation and assert the rejection names that commitment.

- 🟡 **[TY11] Unit — `getReduceData` still lacks the normal surviving-block path.** The direct case covers only reduction to genesis ([`AgreementManager.test.ts#L862-L919`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/unit/AgreementManager.test.ts#L862-L919)).

  > **Fix TY11-FIX**
  >
  > Reduce to a real surviving block and assert its snapshot, encoded state, fork ID, and reduced output.

- 🟡 **[TY12] Unit — `getReduceData` does not cover populated inbound-message bounds and order.** No direct assertion pins the range selected at [`AgreementManager.ts#L409-L414`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/agreementManager/AgreementManager.ts#L409-L414).

  > **Fix TY12-FIX**
  >
  > Stage message blocks before, inside, and after the reduction range; assert only bounded blocks return in application order.

- 🟡 **[TY13] Unit — `getReduceData` does not cover an empty inbound-message range.**

  > **Fix TY13-FIX**
  >
  > Use equal lower and upper hashes with no newly applied messages and assert an empty result.

- 🟡 **[TY14] Unit — `getReduceData` does not cover a missing genesis snapshot.** The genesis branch has a distinct fork-specific error ([`AgreementManager.ts#L375-L385`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/agreementManager/AgreementManager.ts#L375-L385)).

  > **Fix TY14-FIX**
  >
  > Remove the fork's genesis snapshot, invoke genesis reduction, and assert the fork-specific error.

- 🟡 **[TY15] Unit — `getReduceData` does not cover a missing surviving-block snapshot.** This is distinct from missing genesis ([`AgreementManager.ts#L386-L397`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/agreementManager/AgreementManager.ts#L386-L397)).

  > **Fix TY15-FIX**
  >
  > Remove the surviving block's snapshot and assert the missing-latest-snapshot error.

- 🟡 **[TY16] Unit — `getReduceData` does not cover a missing encoded state-machine state.** A valid snapshot can still point at absent local state ([`AgreementManager.ts#L399-L407`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/agreementManager/AgreementManager.ts#L399-L407)).

  > **Fix TY16-FIX**
  >
  > Keep the selected snapshot, remove its encoded state, and assert the missing-state error.

- 🟡 **[TY17] Unit — malformed requested-height boundaries are not exercised.** Production now rejects negative, fractional, non-finite, and unsafe heights before chain reads ([`SpectateService.ts#L457-L469`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/rpc/services/spectate/SpectateService.ts#L457-L469)), but the suite leaves this as a skipped placeholder ([`AgreementManager.test.ts#L140-L144`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/unit/AgreementManager.test.ts#L140-L144)).

  > **Fix TY17-FIX**
  >
  > Exercise the public spectate endpoint with `-1`, a fraction, `NaN`/infinity, and an unsafe integer; assert rejection/blacklisting and no chain read.

- 🟡 **[TY18] Unit — the new synchronized-window owner has no direct no-op fast-path test.** Existing tests stage missing commitments; none proves that a complete local window returns in order without querying historical logs ([`EventSyncService.ts#L255-L263`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/stateManager/EventSyncService.ts#L255-L263)).

  > **Fix TY18-FIX**
  >
  > Use a complete real local window, call `loadSynchronizedWindowCommitments`, assert order, and assert no replay scan was scheduled.

- 🟡 **[TY19] Unit — recovery exhaustion is untested.** After three widening scans, the owner throws with the unresolved commitments ([`EventSyncService.ts#L276-L302`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/stateManager/EventSyncService.ts#L276-L302)); both new consumer tests cover only successful recovery.

  > **Fix TY19-FIX**
  >
  > Keep a real commitment unavailable across all three scans and assert the public consumer rejects with that commitment and does not continue to reduction/payload assembly.

## ♻️ Code reuse

- 🟢 **[CO1] Addressed: synchronized dispute-window loading has one owner.** [`EventSyncService.loadSynchronizedWindowCommitments`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/src/stateManager/EventSyncService.ts#L242-L252) owns commitment retrieval and recovery; reduction and spectator code delegate to it.

  > **Fix CO1-FIX**
  >
  > No change required beyond SO1's channel ownership correction.

## 📏 AGENTS.md adherence

- 🟠 **[AO1] The required E2E interaction is still incomplete.** Historical exact-height synchronization across a participant change is now covered ([`E2E-Spectate.test.ts#L1140-L1282`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/e2e/E2E-Spectate.test.ts#L1140-L1282)), but the spectator recovery test does not cross the peer RPC/request-response boundary described in TO6.

  > **Fix AO1-FIX**
  >
  > Complete the one representative peer-to-peer recovery workflow from TO6.

- 🟡 **[AY2] Two primitive collections still lack domain meaning.** [`confirmationSignerAddresses: string[]`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/fixtures/customRpc/harnessControl/services/query/QueryRpcMethods.ts#L24-L36) lacks an address type or immediate explanation, and [`heightsSeen`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/unit/AgreementManager.test.ts#L451-L464) is `Set<number>` despite the existing `BlockHeight` type.

  > **Fix AY2-FIX**
  >
  > Use `Address[]` and `Set<BlockHeight>`, or add a precise immediate comment where runtime-port serialization requires primitives.

## 🧹 Miscellaneous (scope)

- 🟡 **[MY1] The canonical full gate is not green, although the failure is outside this diff.** `yarn test:parallel` produced 607 passing and 1 failing in `run-510`; the exact case failed again alone in `run-511` at [`invalidStateProofGenesisLinkage.test.ts#L11-L28`](https://github.com/peer3to/state-channels-plus/blob/88167fcbce31f51f7ccf8bf798bd62e7394c8fde/test/e2e/disputeValidation/invalidStateProofGenesisLinkage.test.ts#L11-L28), waiting for peer 1's `onInitiatingDispute`. The same timeout is present in several older persisted runs, including `run-75`, `run-76`, and `run-297`, and this PR does not modify that test or timeout-dispute path. [Full-run failure log](../../../logs/run-510/error_invalidStateProofGenesisLinkage.test__E2E_dispute_validation_DisputeInvalidStateProof_genesis_linkage__unlinked_genesisStateSnapshotData_against_a_valid_genesis_dispute_submitter_slashed_honest_disputer_survives.ansi).

  > **Fix MY1-FIX**
  >
  > Stabilize or repair the pre-existing timeout-dispute setup, then rerun the canonical gate. Do not attribute it to this PR without evidence from the untouched path.

## ❓ Open questions

- 🟢 **[OG1] No product decision is required.** Exact-height proofs already have a clear inclusive contract, and spectator requests must be bound to the state manager's own channel.

  > **Fix OG1-FIX**
  >
  > Apply SO1-FIX and the remaining test fixes without changing protocol semantics.

**Comment by daiagi** [comment]
@lukastanisic99 ready for review

**Review by lukastanisic99 — APPROVED** [review]
